### PR TITLE
Exploration support of symbolic regex into symbolic NFA

### PR DESCRIPTION
**Pull request by veanes**
Added the missing piece intended later for code generation that calls into `MkDerivative()` to create an NFA under `Compiled` flag. 
- Added `SymbolicNFA` that encapsulates the closure of running `MkDerivative()` and `SymbolicRegexeNode.Explore(k)` that creates a `SymbolicNFA` with roughly up to `k` states.
- This creates an NFA whose number of states roughly equals the number of nodes of an AST 
of the regex whose counters have been unwound.  It can be used to validate that 
the underlying NFA is secure (e.g. stays within a `k =1000` state bound). 
-Updated DGML generator to use the new method for exploration and visual inspection.
- Updates some related unit tests.

### CODE REVIEW

**Comment by stephentoub** [comment]
(Please hold off on merging this.  I'm trying to get up a PR that merges the latest from dotnet/runtime main into feature/regexsrm, and it's going to have conflicts with pretty much anything that touches tests.  I'd like to get that in first if that's ok.)

**Comment by veanes** [comment]
No problem. Will hold off.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text.RegularExpressions.Symbolic
+{
+    /// <summary>Represents the exploration of a symbolic regex as a symbolic NFA</summary>
+    internal class SymbolicNFA<S> where S : notnull
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
    internal sealed class SymbolicNFA<S> where S : notnull
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+    {
+        private IBooleanAlgebra<S> _solver;
+        private Transition[] _transitionFunction;
+        private SymbolicRegexNode<S>[] _finalCondition;
+        private HashSet<int> _unexplored;
+        private SymbolicRegexNode<S>[] _nodes;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
        private readonly IBooleanAlgebra<S> _solver;
        private readonly Transition[] _transitionFunction;
        private readonly SymbolicRegexNode<S>[] _finalCondition;
        private readonly HashSet<int> _unexplored;
        private readonly SymbolicRegexNode<S>[] _nodes;
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+        private const int s_deadend_state = -1;
+        private const int s_unexplored_state = -2;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
        private const int DeadendState = -1;
        private const int UnexploredState = -2;
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+
+        /// <summary>If true then some states have not been explored</summary>
+        public bool IsIncomplete => _unexplored.Count > 0;
+
+        private SymbolicNFA(IBooleanAlgebra<S> solver,
+            Transition[] transitionFunction,
+            HashSet<int> unexplored,
+            SymbolicRegexNode<S>[] nodes)
+        {
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
        private SymbolicNFA(IBooleanAlgebra<S> solver, Transition[] transitionFunction, HashSet<int> unexplored, SymbolicRegexNode<S>[] nodes)
        {
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+            Debug.Assert(transitionFunction.Length > 0 && nodes.Length == transitionFunction.Length);
+            _solver = solver;
+            _transitionFunction = transitionFunction;
+            _finalCondition = Array.ConvertAll(nodes, node => node.ExtractNullabilityTest());
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
            _finalCondition = new SymbolicRegexNode<S>[nodes.Length];
            for (int i = 0; i < nodes.Length; i++)
            {
                _finalCondition[i] = nodes[i].ExtractNullabilityTest();
            }
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+                        Debug.Assert(transition._kind == TransitionRegexKind.Lookaround && transition._look is not null && transition._first is not null && transition._second is not null);
+                        // Branch according to nullability of the lookaround condition in the given context
+                        if (transition._look.IsNullableFor(context))
+                        {
+                            transition = transition._first;
+                        }
+                        else
+                        {
+                            transition = transition._second;
+                        }
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
                        // Branch according to nullability of the lookaround condition in the given context
                        transition = transition._look.IsNullableFor(context) ?
                            transition._first :
                            transition._second;
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+            }
+
+            // Continue operating in a mode where several target states can be yielded
+            Debug.Assert(transition._first is not null && transition._second is not null);
+            Stack<Transition> todo = new();
+            todo.Push(transition._second);
+            todo.Push(transition._first);
+            while (todo.Count > 0)
+            {
+                Transition top = todo.Pop();
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
            while (todo.TryPop(out Transition? top))
            {
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+                        Debug.Assert(transition._look is not null && transition._first is not null && transition._second is not null);
+                        // Branch according to nullability of the lookaround condition in the given context
+                        if (transition._look.IsNullableFor(context))
+                        {
+                            todo.Push(transition._first);
+                        }
+                        else
+                        {
+                            todo.Push(transition._second);
+                        }
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
                         todo.Push(transition._look.IsNullableFor(context) ? transition._first : transition._second);
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+                        break;
+                    default:
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs:152]:
```suggestion

                    default:
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+            {
+                return transition;
+            }
+
+            Stack<(TransitionRegex<S>, bool)> work = new();
+            work.Push((tregex, false));
+
+            while (work.Count > 0)
+            {
+                (TransitionRegex<S>, bool) top = work.Pop();
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
            while (work.TryPop(out (TransitionRegex<S>, bool) top)
            {
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+                        // Omit any path that leads to a deadend or is unexplored
+                        if (_leaf >= 0)
+                        {
+                            yield return (pathCondition, null, _leaf);
+                        }
+                        break;
+
+                    case TransitionRegexKind.Union:
+                        Debug.Assert(_first is not null && _second is not null);
+                        foreach (var path in _first.EnumeratePaths(solver, pathCondition))
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]: Nit: for all of these `var`s, can we use the actual type?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+        {
+            Debug.Assert(transitionFunction.Length > 0 && nodes.Length == transitionFunction.Length);
+            _solver = solver;
+            _transitionFunction = transitionFunction;
+            _finalCondition = Array.ConvertAll(nodes, node => node.ExtractNullabilityTest());
+            _unexplored = unexplored;
+            _nodes = nodes;
+        }
+
+        /// <summary>Total number of states, 0 is the ininital state, states are numbered from 0 to StateCount-1</summary>
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
        /// <summary>Total number of states, 0 is the initial state, states are numbered from 0 to StateCount-1</summary>
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+
+        /// <summary>Returns true if the state is nullable for all contexts</summary>
+        public bool IsNullable(int state) => _finalCondition[state].IsNullable;
+
+        /// <summary>Gets the underlying node of the state</summary>
+        public SymbolicRegexNode<S> GetNode(int state) => _nodes[state];
+
+        /// <summary>Enumerates all target states from the given source state</summary>
+        /// <param name="sourceState">must be a an integer between 0 and StateCount-1</param>
+        /// <param name="input">must be a value that acts as a minterm for the transitions eminating from the source state</param>
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
        /// <param name="input">must be a value that acts as a minterm for the transitions emanating from the source state</param>
```
?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+            workState.front.Push(0);
+
+            Dictionary<int, Transition> transitions = new();
+            Stack<int> front = new();
+
+            while (workState.front.Count > 0)
+            {
+                Debug.Assert(front.Count == 0);
+
+                // Work Breath-First in layers, swap front with workState.front
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]:
```suggestion
                // Work Breadth-First in layers, swap front with workState.front
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs (hunk 11 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicNFA.cs]: ok, will do

**Comment by veanes** [comment]
I rebased the branch to the latest feature/regexsrm and squashed/simplified the edit history a bit.

**Comment by veanes** [comment]
Merging at this time. Unrelated test System.IO.Tests.Enumeration.SpecialDirectoryTests_DirectoryInfo_GetDirectories.SkippingHiddenFiles has failed.

**Pull request merged**